domain/src/main/java/org/oppia/android/domain/question/QuestionAssessmentCalculation.kt

```diff
@@ -0,0 +1,118 @@
+package org.oppia.android.domain.question
+
+import org.oppia.android.app.model.FractionGrade
+import org.oppia.android.app.model.UserAssessmentPerformance
+import javax.inject.Inject
+import kotlin.math.max
+
+/**
+ * Private class that computes the state of the user's performance at the end of a practice session.
+ * This class can exist across multiple training session instances, but calling code is responsible
+ * for ensuring it is properly reset. This class is not thread-safe, so it is the calling code's
+ * responsibility to synchronize access to this class.
+ */
+internal class QuestionAssessmentCalculation constructor(
+  private val viewHintPenalty: Int,
+  private val wrongAnswerPenalty: Int,
+  private val maxScorePerQuestion: Int,
+  private val questionSessionMetrics: List<QuestionSessionMetrics>,
+  private val totalScore: MutableFractionGrade,
+  private val scorePerSkillMapping: MutableMap<String, MutableFractionGrade>
+) {
+  /** Calculate the user's overall score and score per skill for this practice session. */
+  private fun calculateScores() {
+    for (questionMetric in questionSessionMetrics) {
+      val totalHintsPenalty =
+        questionMetric.numberOfHintsUsed * viewHintPenalty
+      val totalWrongAnswerPenalty =
+        max((questionMetric.numberOfAnswersSubmitted - 1), 0) * wrongAnswerPenalty
+      val questionScore =
+        if (questionMetric.didViewSolution) 0
+        else max(maxScorePerQuestion - totalHintsPenalty - totalWrongAnswerPenalty, 0)
+      this.totalScore.pointsReceived += questionScore
+      this.totalScore.totalPointsAvailable += maxScorePerQuestion
+      for (linkedSkillId in questionMetric.question.linkedSkillIdsList) {
+        scorePerSkillMapping[linkedSkillId]?.let {
+          it.pointsReceived += questionScore
+          it.totalPointsAvailable += maxScorePerQuestion
+        }
+      }
+    }
+  }
+
+  /** Compute the overall score as well as the score and mastery per skill. */
+  internal fun computeAll(): UserAssessmentPerformance {
+    calculateScores()
+    // TODO(#3067): Create mastery calculations method
+
+    // Set up the return values
+    val finalScore = FractionGrade.newBuilder().apply {
+      pointsReceived = totalScore.pointsReceived.toDouble() / maxScorePerQuestion
+      totalPointsAvailable = totalScore.totalPointsAvailable.toDouble() / maxScorePerQuestion
+    }.build()
+
+    val finalScorePerSkillMapping = mutableMapOf<String, FractionGrade>()
+    for (score in scorePerSkillMapping) {
+      if (score.value.totalPointsAvailable != 0) { // Exclude entries with denominator of 0
+        finalScorePerSkillMapping[score.key] = FractionGrade.newBuilder().apply {
+          pointsReceived = score.value.pointsReceived.toDouble() / maxScorePerQuestion
+          totalPointsAvailable = score.value.totalPointsAvailable.toDouble() / maxScorePerQuestion
+        }.build()
+      }
+    }
```

Comment:
```suggestion
    val finalScorePerSkillMapping = skillPerScoreMapping.filter { (_, fractionGrade) ->
      fractionGrade.totalPointsAvailable != 0
    }.mapValues { (_, fractionGrade) ->
      FractionGrade.newBuilder().apply {
        pointsReceived = fractionGrade.pointsReceived.toDouble() / maxScorePerQuestion
        totalPointsAvailable = fractionGrade.totalPointsAvailable.toDouble() / maxScorePerQuestion
      }.build()
    }
```

This is more Kotlin-idiomatic, and avoids using mutable state at all. It can also be made more efficient if you use sequences, instead, but the iterable version is fine.

Reply:
Thanks for the suggestion! I'll keep this in mind for next time as well